offline appenderator factory.

b-slim opened the pull request:
This patch adds an offline factory to create offline appenderator.

b-slim commented:
@gianm can you please check this. I am not sure about the inject stuff if such class should be only created by Jackson or it is ok to have the `@Inject` annotation

leventov commented on server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java:
```diff
@@ -0,0 +1,155 @@
+    );
+    Assert.assertEquals(null, appenderator.startJob());
+    Assert.assertEquals("dataSourceName", appenderator.getDataSource());
+    appenderator.close();
```
Use try-with-resources instead of explicit `close()`

b-slim replied:
ok

leventov commented on server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java:
```diff
@@ -0,0 +1,155 @@
+    appenderator.close();
+  }
+
+}
```
Miss newline

leventov commented on server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java:
```diff
@@ -0,0 +1,155 @@
+        new FireDepartmentMetrics()
+    );
+    Assert.assertEquals(null, appenderator.startJob());
+    Assert.assertEquals("dataSourceName", appenderator.getDataSource());
```
Maybe try to add some sample row to Appenderator, to see that it works (and cache is non-null)?

leventov commented on server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java:
```diff
@@ -124,8 +124,8 @@ public boolean isAnnounced(DataSegment segment)
         indexMerger,
-        null,
-        null
+        MapCache.create(500000),
```
Doesn't seem to be universal to use specifically `MapCache` (`CaffeineCache` is also an in-memory, local cache) of the specific size. Maybe replace `Appendirators` with something like `AppendiratorBuilder`, which allows to configure those batteries of arguments one-by-one using named configuration methods? It may distinguish between "realtime" and "offline" on different stages, e. g. in the beginning of the chain:
```java
AppenderatorBuilder.offline() // or .realtime(segmentAnnouncer)
  .setIndexMerger(...)
  .setCache(...)
  .build();
```

or in the end:
```java
AppenderatorBuilder.create()
  .setIndexMerger(..)
  .setCache(..)
  .buildRealtime() // or .buildOffline()
```

Or in some other way

leventov replied:
Could you please also test this cache configuration missing, e. g. add something to test that would fail before this change? (Also see my comment to `DefaultOfflineAppenderatorFactoryTest`)

b-slim replied:
the test in this PR will fail if the cache was missing is that not enough ?

b-slim replied:
@leventov thanks for the comments about the cache actually it is not really needed i will remove it.

b-slim commented:
Thanks @leventov, i have addressed the comments minus the one about the builder. You think a Builder is really better ? if so i can change it. Please let me know.

leventov commented:
@b-slim let some other Druid committer decide.

b-slim commented:
@leventov thanks for the feedback !

b-slim commented:
@gianm can you take look at this ? it is pretty minor

cheddar commented on server/src/main/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactory.java:
```diff
@@ -0,0 +1,60 @@
+  private final IndexIO indexIO;
+  private final IndexMerger indexMerger;
+
+  @Inject
```
Given that we expect it to be instantiated by Jackson, this would need to be `@JsonCreate`

nishantmonu51 commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -912,7 +912,9 @@ public Object apply(@Nullable Object input)
             );
             for (FireHydrant hydrant : sink) {
-              cache.close(SinkQuerySegmentWalker.makeHydrantCacheIdentifier(hydrant));
+              if (cache != null) {
```
when can this be null ?

nishantmonu51 replied:
We should also remove the null check from the constr also then?

nishantmonu51 commented on server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java:
```diff
@@ -129,5 +128,4 @@ public boolean isAnnounced(DataSegment segment)
     );
   }
 
-
```
unrelated change to this PR.

b-slim replied:
this is part of bug fix. when using offline appenderator we don't server queries so no need for cache.

b-slim replied:
it is just less empty line is that not ok ?

nishantmonu51 replied:
it's ok.

cheddar commented:
Btw, you should be able to instantiate one of these locally on your dev box, which should make it clear if the injection/jackson stuff is working.

It should also be unit testable...

b-slim commented:
@cheddar thanks i guess the new UT fix that.

b-slim commented:
@cheddar i have addressed all the comments Thanks !

cheddar commented:
👍
